### DISCUSSION TITLE
Perform  3D_6000_gr.i unable to initialize the system

### DISCUSSION THREAD

**Discussion by Pengwei007** [idaholab/moose#17198] [Q&A Modules: Phase field] [answered]
Hello MOOSE users,
Accept my warm greetings. I used the input file [3D_6000_gr.i](https://github.com/idaholab/moose/blob/next/modules/phase_field/examples/grain_growth/3D_6000_gr.i) that already existed in the grain growth, but an error was displayed when the input file was executed. I hope someone can solve my problem. The commands I run and the information displayed in the terminal are as follows,
`mpirun -np 15 ~/projects/panda/panda-opt -i 3D_6000_gr.i`
```
Building mesh ............................................................................
.....................................                                                      [114.32 s]
Caching mesh information ......................                                            [ 23.97 s]
Initializing equation system ..................................
===================================================================================
=   BAD TERMINATION OF ONE OF YOUR APPLICATION PROCESSES
=   PID 2514893 RUNNING AT haochenmd-PowerEdge-T640
=   EXIT CODE: 9
=   CLEANING UP REMAINING PROCESSES
=   YOU CAN IGNORE THE BELOW CLEANUP MESSAGES
===================================================================================
YOUR APPLICATION TERMINATED WITH THE EXIT STRING: Killed (signal 9)
This typically refers to a problem with your application.
Please see the FAQ page for debugging suggestions
```

Best Regards,
Wordsworth

**Comment by permcody** [comment]
You are out of memory. This input file is rather old and it used "ReplicatedMesh" rather than the newer "DistributedMesh" capability. As a result you have to have on the order of 8-10GB per MPI rank, which almost no clusters have. I ran this by using only a few ranks per node on the cluster at the time (again about a quarter of what was available).

The movie from that run is in the MOOSE gallery ([MOOSE Gallery](https://mooseframework.inl.gov/gallery.html) - 3D Grain Tracking). If you need to run it as is, you'll have to reduce the number of CPUs per node. An alternative would be to try using distributed mesh but you may have more work to get it fully adapted. Are you trying to expand on this capability or just trying things out?

**Reply by Pengwei007** [reply]
Thank you so much for your suggestion. The purpose of this example is to take a look at the three-dimensional ideal grain growth. Later, I may simulate the three-dimensional grain growth based on this model.
According to your suggestion, there are two ways to modify it.
First, instead of using 15 Processors to compute it in parallel, it may need to be reduced to 5 or less, right?
Second, we can change the grid type from replicated grid to distributed grid to reduce memory overflow. Is my understanding correct?
And when I add `parallel_type = distributed #` in [mesh] block included in the input file, and then run the command line in the terminal` mpirun -np 5 ~/projects/panda/panda-opt -i 3D_6000_gr.i `
But it still showing a similar error above. So if I using distributed mesh, What do I need to consider?

**Reply by GiudGiud** [reply]
Hi 

You may still be running out of memory. 
Can you monitor your memory, using `top` or `htop` and see if that's the problem?

Guillaume

**Reply by Pengwei007** [reply]
Dear @GiudGiud,

This is the running status of the computer after I run [3D_6000_gr.i](https://github.com/WuHao00-C/dragon/blob/main/3D_6000/3D_6000_gr.i). It shows that the memory is not exceeded. This time I used 30 cores for calculation.
![图片1](https://user-images.githubusercontent.com/79966177/114328741-0d7efa80-9b70-11eb-92ea-44bc28fde8ea.png)

In addition, the information output to the terminal outputs the following information. I don't know which parameter in the input should be modified to solve this problem. Would you have any good suggestions?

> Grain #151 intersects Grain #179 (variable index: 0)
> - Depth 0: Remapping grain #151 from variable index 0 to 18 which currently contains zero grains.
> 
> Grain #155 intersects Grain #180 (variable index: 0)
> - Depth 0: Remapping grain #155 from variable index 0 to 17 which currently contains zero grains.
> 
> Grain #157 intersects Grain #186 (variable index: 0)
> - Depth 0: Remapping grain #157 from variable index 0 to 16 which currently contains zero grains.
> 
> Grain #164 intersects Grain #165 (variable index: 0)
> - Depth 0: Remapping grain #164 from variable index 0 to 24 whose closest grain (#107) is at a distance of 75.1332
> 
> Grain #191 intersects Grain #289 (variable index: 0)
> - Depth 0: Remapping grain #191 from variable index 0 to 25 whose closest grain (#80) is at a distance of 98.4886
> 
> Grain #193 intersects Grain #451 (variable index: 0)
> - Depth 0: Remapping grain #193 from variable index 0 to 20 whose closest grain (#54) is at a distance of 83.9226

The detailed output information is in the [3D_6000_gr.log](https://github.com/WuHao00-C/dragon/blob/main/3D_6000/3D_6000_gr.log) file.

Hao

**Reply by GiudGiud** [reply]
Is this `top` screenshot captured right before the crash?
@permcody will know what these warnings mean

Your simulation got a lot further than initially it seems, with more processors, so the distributed mesh helped. Cody if you want to elaborate on what you meant by what's needed to get it fully adapted. 

Guillaume

**Reply by Pengwei007** [reply]
hi @GiudGiud, thank you so much for your feedback.

> Is this top screenshot captured right before the crash?

The above detection using `htop` is the computer usage in the `Projecting initial condition` of this calculation. 

We found that the memory is not overflowing. According to your question, I ran this simulation again, and traced the memory in the "Time Step 1, time = 0.0002, dt = 0.0002" stage. The memory usage accounted for 99.06% of the total memory. It did have a memory overflow. 

How should I modify the input file to prevent memory overflow during the Time Step phase of the calculation phase?

The following are the specific details of this simulation and the input terminal file,

> mpirun -np 30 ~/projects/panda/panda-opt -i 3D_6000_gr.i > 3D_6000_gr_02.log

> Grain #5899 intersects Grain #5911 (variable index: 13)
> - Depth 0: Remapping grain #5899 from variable index 13 to 22 whose closest grain (#487) is at a distance of 27
> 
> Finished inside of GrainTracker
> 
> Outputting nemesis ...                                                                     [  4.63 s]
> 
> Time Step 0, time = 0
> 
> Postprocessor Values:
> +----------------+----------------+----------------+----------------+----------------+----------------+
> | time           | DOFs           | dt             | grain_tracker  | n_elements     | n_nodes        |
> +----------------+----------------+----------------+----------------+----------------+----------------+
> |   0.000000e+00 |   0.000000e+00 |   0.000000e+00 |   6.000000e+03 |   0.000000e+00 |   0.000000e+00 |
> +----------------+----------------+----------------+----------------+----------------+----------------+
> 
> Outputting checkpoint 
> ...............                                                                            [ 16.64 s]
> 
> Time Step 1, time = 0.0002, dt = 0.0002
> Computing initial residual ...........................................                     [ 44.71 s]
>  0 Nonlinear |R| = 1.958070e+07
> 
> 
> =   BAD TERMINATION OF ONE OF YOUR APPLICATION PROCESSES
> =   PID 2174924 RUNNING AT pwmoose-PowerEdge-T640
> =   EXIT CODE: 9
> =   CLEANING UP REMAINING PROCESSES
> =   YOU CAN IGNORE THE BELOW CLEANUP MESSAGES
> 
> YOUR APPLICATION TERMINATED WITH THE EXIT STRING: Killed (signal 9)
> This typically refers to a problem with your application.
> Please see the FAQ page for debugging suggestions

Best Regards,
hao

**Reply by GiudGiud** [reply]
Hello

You could try using a different time stepping scheme, bdf2 requires storing 3 statepoints, some first order schemes will only store 2. 

Otherwise you could;
- profile the code and try to optimize it for memory. This might be more involved than you want 
- run a smaller case. Maybe 5000 grains could run?
- use a computing cluster. Most nodes these days have at least 128 GB of memory, so you could have more than your current machine in just 3 nodes. 

Guillaume

**Reply by Pengwei007** [reply]
Thank you @GiudGiud  very much for your careful guidance. I reduced the number of grains to 100. When the simulation reached the second step, there was still a memory overflow.

This is the result of the input file I used and output to the terminal,
[3D_6000_gr_07.i](https://github.com/WuHao00-C/dragon/blob/main/3D_6000/3D_6000_gr_07.i)
[3D_6000_gr_07.log](https://github.com/WuHao00-C/dragon/blob/main/3D_6000/3D_6000_gr_07.log)

Similarly, I also performed a simulation of 500 grains, and the memory overflowed without finishing the first step. I don't know if I have done stupid things, always making this three-dimensional simulation out of memory.

hao

**Comment by Traiwit** [comment]
interesting, just wonder if changing the PETSc solving options would help to reduce memory usage? I also have the same problem with my simulation (4m elements) on 256gb system, I only can run with mpi up to 7 cores (out of 32 cores) before it's out of memory. @permcody @GiudGiud 

 ```
 type = Transient
  solve_type = 'NEWTON'
  start_time = 0.0
  end_time = 11
  nl_abs_tol = 1e-2
  nl_rel_tol = 1e-2
  # l_max_its = 30
  # nl_max_its = 1000
   petsc_options_iname = '-pc_type -pc_hypre_type'
   petsc_options_value = 'hypre    boomeramg'

```

**Reply by GiudGiud** [reply]
@fdkong will know more about memory consumption of each petsc solver

**Comment by fdkong** [comment, marked as answer]
Please see the discussion at https://github.com/idaholab/moose/discussions/17422#discussioncomment-704849

**Reply by Pengwei007** [reply]
Thank you very much for your reply, @fdkong. 

I read what you quoted by ykvishal, but the  ykvishal‘s calculation example  has not reached the Time Step stage, and is still in the `Projecting initial condition` stage. When I run on my server, the memory overflow occurred during the `Time Step` stage. Even if the number of grains was reduced to 100, it was detected that the memory overflowed temporarily at this stage. 

When I ran a two-dimensional simulation with me, this problem did not occur even with 25600 grains. The input file used was named [poly_grain_growth_2D_eldrforce_25600_fcc_copper.i](https://github.com/WuHao00-C/dragon/blob/main/2D_25600/poly_grain_growth_2D_eldrforce_25600_fcc_copper.i)

Finally, I will read #17422  carefully, looking forward to finding the answer here.

hao
